### ISSUE TITLE
Unable to initialize database access (DBI); rc='-1', error='No such file or directory (2)'

### ISSUE THREAD

**Issue by rschamizo** [syslog-ng/syslog-ng#4033]
# syslog-ng
## Version of syslog-ng
3.36.1

## Platform
NAME="Oracle Linux Server"
VERSION="8.6"

## Debug bundle
[syslog.debun.syslog-ib-1.2022-06-02_08-24.1586.zip](https://github.com/syslog-ng/syslog-ng/files/8822138/syslog.debun.syslog-ib-1.2022-06-02_08-24.1586.zip)

# Issue
## Failure
```
syslog-ng: The -d/--debug option no longer implies -e/--stderr, if you want to redirect internal() source to stderr please also include -e/--stderr option
[2022-06-02T08:38:06.387127] WARNING: the match() filter without the use of the value() option is deprecated and hinders performance, please use a more specific filter like message() and/or program() instead; location='/etc/syslog-ng/syslog-ng.conf:49:33'
[2022-06-02T08:38:06.388245] WARNING: With use-dns(no), dns-cache() will be forced to 'no' too!;
[2022-06-02T08:38:06.389000] Unable to initialize database access (DBI); rc='-1', error='No such file or directory (2)'
[2022-06-02T08:38:06.389018] Error initializing message pipeline; plugin_name='sql', location='/etc/syslog-ng/syslog-ng.conf:37:3'
```

The problem is connecting to Oracle Database. Syslog cannot find the libdbi-drivers. I have installed oracle instantclient, building from source libdbi, and libdbi-drivers (with support for oracle and mysql, the default libraries doesnt come with oracle drivers), adding libraries to /etc/ld.so.conf.d/geoip.conf (And in LD_LIBRARY_PATH just in case). When i check the libdbi-drivers installation, with 'make check', all tests pass. I dont know what i am missing, of it is a bug.

## Steps to reproduce
I started with a clean vagrant box with generic/oracle8 v4.0.0 (Vagrant 2.2.19)
[Vagrantfile.txt](https://github.com/syslog-ng/syslog-ng/files/8822185/Vagrantfile.txt)

### Install database clients
```
sudo su -
yum install oracle-instantclient-release-el8 mysql-devel mariadb.x86_64 mariadb-server.x86_64 mysql-devel
```

Building libdbi drivers demands a ORACLE_HOME, so i installed oracle database xe
(HOST Fedora) scp /home/user/Downloads/oracle-database-xe-21c-1.0-1.ol8.x86_64.rpm syslog-ib-2:/home/vagrant (https://www.oracle.com/es/database/technologies/xe-downloads.html)

`dnf -y localinstall oracle-database-xe-21c-1.0-1.ol8.x86_64.rpm`

### Install build dependencies
(HOST Fedora) scp -r /home/user/Downloads/opensp-1.5.2-28.el8.x86_64.rpm  syslog-ib-2:/home/vagrant (https://www.rpmfind.net/linux/RPM/centos/8-stream/powertools/x86_64/Packages/opensp-1.5.2-28.el8.x86_64.html)

(HOST Fedora) scp -r /home/user/Downloads/openjade-1.3.2-57.el8.x86_64.rpm  syslog-ib-2:/home/vagrant (https://www.rpmfind.net/linux/RPM/centos/8-stream/powertools/x86_64/Packages/openjade-1.3.2-57.el8.x86_64.html)

(HOST Fedora) scp -r /home/user/Downloads/libdbi-drivers-0.9.0.tar.gz  syslog-ib-2:/home/vagrant (https://sourceforge.net/projects/libdbi-drivers/files/libdbi-drivers/)

(HOST Fedora) scp -r /home/user/Downloads/libdbi-0.9.0.tar.gz  syslog-ib-2:/home/vagrant (https://sourceforge.net/projects/libdbi/files/libdbi/)

```
dnf group install "Development Tools"
dnf -y localinstall opensp-1.5.2-28.el8.x86_64.rpm
dnf -y localinstall openjade-1.3.2-57.el8.x86_64.rpm 
tar -xzf libdbi-drivers-0.9.0.tar.gz
tar -xzf libdbi-0.9.0.tar.gz
```

### Build libdbi
```
cd /home/vagrant/libdbi-0.9.0
./configure
make
make install
```

### Build libdbi-drivers
```
echo "/usr/local/lib" > /etc/ld.so.conf.d/geoip.conf
echo "/usr/local/lib/dbd" >>  /etc/ld.so.conf.d/geoip.conf
echo "/opt/oracle/product/21c/dbhomeXE/lib" >>  /etc/ld.so.conf.d/geoip.conf
ldconfig
```

```
echo "ORACLE_HOME=/opt/oracle/product/21c/dbhomeXE" >> /etc/environment
echo "ORACLE_SID=XE" >> /etc/environment
echo "ORAENV_ASK=NO" >> /etc/environment
echo "ORACLE_BASE=/opt/oracle" >> /etc/environment
reboot
sudo su -
```

```
cd /home/vagrant/libdbi-drivers-0.9.0/
./configure --with-oracle --with-mysql 
make
make install
make check
```

### Now install syslog-ng
```
wget https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm
rpm -Uvh epel-release-latest-8.noarch.rpm

cd /etc/yum.repos.d/
wget https://copr.fedorainfracloud.org/coprs/czanik/syslog-ng336/repo/epel-8/czanik-syslog-ng335-epel-8.repo
yum install syslog-ng syslog-ng-sql
```
(PASTE SYSLOG CONF)

```
cat << EOF > /etc/tnsnames.ora
LOGS =
(DESCRIPTION =
(ADDRESS_LIST =
(ADDRESS = (PROTOCOL = TCP)
(HOST = 10.0.2.2)
(PORT = 1539))
)
(CONNECT_DATA =
(SERVICE_NAME = x2)
)
)
EOF
```

```
sudo systemctl stop syslog-ng
/usr/sbin/syslog-ng -d -F -p /var/run/syslogd.pid
```

## Configuration
[syslog.txt](https://github.com/syslog-ng/syslog-ng/files/8822518/syslog.txt)

The Database connection no is necessary in this case, because the real problem is syslog finding libdbi-drivers, not the connection itself.

Thanks you if you can help me!

**Comment by rschamizo** [comment]
It works like a charm!

I dont understand by syslog-ng doesnt get the libraries path of /etc/ld.so.conf.d/geoip.conf, but now works!

![image](https://user-images.githubusercontent.com/78882304/172914933-d9fae41e-6a88-409a-a4fb-f8aaab2ebd05.png)

Thank you very much!

**Comment by gaborznagy** [comment]
I think the reason is that libdbi always uses a directory path when loads drivers with dlopen.

Libdbi can be initialized with a driver directory by the application (we don't do that), or it uses it's own default driver location (which can be changed build time as @OverOrion suggested), but it never allows dlopen to search on the paths used by the dynamic linker.

**Comment by bazsi** [comment]
This seems to have been resolved. Closing.

**Comment by bazsi** [comment]
Probably the libdbi driver for Oracle has not been built. Try running
syslog-ng with strace and see where the ENOENT error is generated.

When I was at Balabit (now One Identity), we supported Oracle but we had
some patches on libdbi-drivers. We have published the patched version on
GitHub, eg
https://github.com/balabit/libdbi-drivers/commits/master

Maybe that has some hints.

Also, I am not sure it's such a good idea to store logs to SQL these days.

There are so many other tools that are better. (Splunk, Elastic, Loki or
something).

Bazsi
